### ISSUE TITLE
StopwatchNode should move to front when its buttons are pressed.

### ISSUE THREAD

**Issue by samreid** [phetsims/scenery-phet#659]
From https://github.com/phetsims/wave-interference/issues/450.  @KatieWoe said:

> I noticed in 1.1.0 rc.3 that this works in most cases, but if the stopwatch is behind another tool, and you press one of its buttons, it does not jump to the top. It does if you grab a non-button area.
![moveforwardbuttons](https://user-images.githubusercontent.com/41024075/104382389-3c42f180-54eb-11eb-89b3-c06c858c5278.gif)

I replied:

> There is a natural place we could add a `moveToFront` listener in the restart button.  But the play pause button is a `BooleanRectangularToggleButton` and has no apparent places to listen for press.  We could listen to the play/pause property and move to front when it changes, but then it would also move to front when it changes without a button press.  @jonathanolson and @pixelzoom I noticed you have commits in StopwatchNode, can you please recommend how to proceed?

**Comment by samreid** [comment]
@jonathanolson said:

What is the desired behavior? I'm still very uneasy about having a default "moveToFront" inside a component (but I guess you can always wrap it with a Node if desired?)

> But the play pause button is a BooleanRectangularToggleButton and has no apparent places to listen for press.

That seems like something that might be nice to have for this and other reasons.

I think the better approach is to NOT have moveToFront baked into StopwatchNode (especially just in the dragTarget!!!), but instead a client can add a listener to it that moves it to the front when interacted with.

**Comment by samreid** [comment]
I discussed with @arouinfar and we agreed that it's not a blocking publication issue that clicking the buttons doesn't bring the stopwatch to the front.  Since waves intro 1.1.0-rc.3 moves the stopwatch node to front for all except for the buttons, I'll remove this issue from the milestone.

> I think the better approach is to NOT have moveToFront baked into StopwatchNode (especially just in the dragTarget!!!), but instead a client can add a listener to it that moves it to the front when interacted with.

That seems like a nontrivial issue which would probably need to be discussed with the dev team, and may go against parts of the decision in https://github.com/phetsims/scenery-phet/issues/608#issuecomment-663190147.  Up to @jonathanolson if we should open an issue for that.

**Comment by samreid** [comment]
@jonathanolson said:

> Or, presumably we'd want all StopwatchNodes to moveToFront with the buttons, so StopwatchNode could create a listener on ALL of itself?

**Comment by samreid** [comment]
This patch seems to be working OK, is it what you had in mind?

```diff
Index: js/StopwatchNode.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/StopwatchNode.js b/js/StopwatchNode.js
--- a/js/StopwatchNode.js	(revision 18007999e489b1f4709ac13fb18695329d80b9d4)
+++ b/js/StopwatchNode.js	(date 1611963761694)
@@ -159,7 +159,7 @@
     const stopwatchVisibleListener = visible => {
       this.visible = visible;
       if ( visible ) {
-        this.moveToFront();
+        // this.moveToFront();
       }
       else {
 
@@ -205,7 +205,7 @@
       // The StopwatchNode always calls moveToFront on drag start
       const startOption = dragListenerOptions.start;
       dragListenerOptions.start = () => {
-        this.moveToFront();
+        // this.moveToFront();
         startOption && startOption();
       };
 
@@ -238,6 +238,10 @@
     // @private
     this.numberDisplay = numberDisplay;
 
+    this.addInputListener( {
+      down: () => this.moveToFront()
+    } );
+
     // support for binder documentation, stripped out in builds and only runs when ?binder is specified
     assert && phet.chipper.queryParameters.binder && InstanceRegistry.registerDataURL( 'scenery-phet', 'StopwatchNode', this );
   }
@@ -293,7 +297,7 @@
    * @param {Object} [options]
    * @returns {function(time:number):string} - see NumberDisplay options.numberFormatter
    */
-  static createRichTextNumberFormatter( options ){
+  static createRichTextNumberFormatter( options ) {
 
     options = merge( {
 
```

**Comment by samreid** [comment]
I noticed the patch doesn't pop it to the front when dragging out of the toolbox though.

**Comment by samreid** [comment]
@jonathanolson said:

> I noticed the patch doesn't pop it to the front when dragging out of the toolbox though.

Presumably we'd want to handle touch-snag and the toolbox pattern! Having a DragListener with `attach:false` would handle the first, however the second would almost require information from the actual DragListener being triggered, right?

**Comment by samreid** [comment]
Leaving unassigned since this was determined unnecessary for https://github.com/phetsims/wave-interference/issues/450

**Comment by pixelzoom** [comment]
Imo, I don't think `moveToFront` (or any layering decision) belongs in a reusable UI component.   That is context-specific behavior, and its certainly not consistent across interactive UI components.   If it were my call, I'd remove the calls to `moveToFront` that occur when dragging starts, and when visiblity changes.  Those should be the responsibility of the client.

**Comment by pixelzoom** [comment]
Here's an example of why I don't believe this functionality belongs in StopwatchNode.  The _Lab_ screen in Pendulum Lab has 3 draggable tools - Ruler, Stopwatch, Period Timer - shown in the screenshot below.  

<img width="1122" alt="screenshot_880" src="https://user-images.githubusercontent.com/3046552/108880468-8e356780-75bf-11eb-8198-0f49bed7e42e.png">

Only one of these tools, the Stopwatch, does `moveToFront`.  This is inconsistent behavior.  And it's not a matter of adding `moveToFront` to the other common-code components.  RulerNode, for example, is not interactive -- making it interacrive (e.g. draggable) is up to the client. 

My recommendation is:

* Remove `moveToFront` from StopwatchNode.
* Create `MoveToFrontListener extends PressListener`, and have clients add an instance of that listener to UI components (including StopwatchNode) that should move to front when pressed.

There are currently 13 occurrences of `import StopwatchNode` in sims.

@samreid thoughts?

**Comment by samreid** [comment]
That sounds like a reasonable proposal.  I implemented it and tested in Wave Interference for the stopwatch:

<details>

```diff

Index: main/scenery/js/listeners/MoveToFrontListener.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/scenery/js/listeners/MoveToFrontListener.js b/main/scenery/js/listeners/MoveToFrontListener.js
new file mode 100644
--- /dev/null	(date 1615386795403)
+++ b/main/scenery/js/listeners/MoveToFrontListener.js	(date 1615386795403)
@@ -0,0 +1,30 @@
+// Copyright 2021, University of Colorado Boulder
+import merge from '../../../phet-core/js/merge.js';
+import scenery from '../scenery.js';
+import PressListener from './PressListener.js';
+
+/**
+ * Moves the target node to the front on press.  Note that if the object is dragged from a toolbox, you will likely
+ * need to add a moveToFront in that code path separately.
+ *
+ * @author Sam Reid (PhET Interactive Simulations)
+ */
+class MoveToFrontListener extends PressListener {
+
+  /**
+   * @param {Node} targetNode - the node that should be moved to the front.
+   * @param {Object} [options]
+   */
+  constructor( targetNode, options ) {
+    assert && assert( !options || options.hasOwnProperty( 'press' ), 'MoveToFrontListener supplies press' );
+
+    options = merge( {}, {
+      press: () => targetNode.moveToFront()
+    } );
+    super( options );
+  }
+}
+
+scenery.register( 'MoveToFrontListener', MoveToFrontListener );
+
+export default MoveToFrontListener;
\ No newline at end of file
Index: main/wave-interference/js/common/view/ToolboxPanel.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/wave-interference/js/common/view/ToolboxPanel.js b/main/wave-interference/js/common/view/ToolboxPanel.js
--- a/main/wave-interference/js/common/view/ToolboxPanel.js	(revision 2aefebb2e0f5da15a28017a9600d7d1f3d3d85dc)
+++ b/main/wave-interference/js/common/view/ToolboxPanel.js	(date 1615386555593)
@@ -57,6 +57,7 @@
     // The draggable icon, which has an overlay to make the buttons draggable instead of pressable
     initializeIcon( stopwatchNodeIcon, isStopwatchVisibleProperty, event => {
       stopwatchNode.center = this.globalToParentPoint( event.pointer.point );
+      stopwatchNode.moveToFront();
 
       // stopwatchNode provided as targetNode in the DragListener constructor, so this press will target it
       stopwatchNode.dragListener.press( event );
Index: main/scenery-phet/js/StopwatchNode.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/scenery-phet/js/StopwatchNode.js b/main/scenery-phet/js/StopwatchNode.js
--- a/main/scenery-phet/js/StopwatchNode.js	(revision 54d0b885b32d396d8b84197b8946a02c1161d0f3)
+++ b/main/scenery-phet/js/StopwatchNode.js	(date 1615383837233)
@@ -158,10 +158,7 @@
 
     const stopwatchVisibleListener = visible => {
       this.visible = visible;
-      if ( visible ) {
-        this.moveToFront();
-      }
-      else {
+      if ( !visible ) {
 
         // interrupt user interactions when the stopwatch is made invisible
         this.interruptSubtreeInput();
@@ -199,13 +196,13 @@
         targetNode: this,
         positionProperty: stopwatch.positionProperty,
         dragBoundsProperty: dragBoundsProperty,
+        attach: false, // allow other listeners to process, such as moveToFront()
         tandem: options.tandem.createTandem( 'dragListener' )
       }, options.dragListenerOptions );
 
       // The StopwatchNode always calls moveToFront on drag start
       const startOption = dragListenerOptions.start;
       dragListenerOptions.start = () => {
-        this.moveToFront();
         startOption && startOption();
       };
 
Index: main/wave-interference/js/common/view/WaveInterferenceStopwatchNode.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/wave-interference/js/common/view/WaveInterferenceStopwatchNode.js b/main/wave-interference/js/common/view/WaveInterferenceStopwatchNode.js
--- a/main/wave-interference/js/common/view/WaveInterferenceStopwatchNode.js	(revision 2aefebb2e0f5da15a28017a9600d7d1f3d3d85dc)
+++ b/main/wave-interference/js/common/view/WaveInterferenceStopwatchNode.js	(date 1615386384261)
@@ -10,6 +10,7 @@
 import merge from '../../../../phet-core/js/merge.js';
 import Range from '../../../../dot/js/Range.js';
 import StopwatchNode from '../../../../scenery-phet/js/StopwatchNode.js';
+import MoveToFrontListener from '../../../../scenery/js/listeners/MoveToFrontListener.js';
 import waveInterference from '../../waveInterference.js';
 import WaveInterferenceConstants from '../WaveInterferenceConstants.js';
 import WaveInterferenceText from './WaveInterferenceText.js';
@@ -52,6 +53,8 @@
       unitsProperty.value = scene.timeUnits;
       this.redrawNumberDisplay();
     } );
+
+    this.dragTarget.addInputListener( new MoveToFrontListener( this ) );
   }
 }
 

```

</details>

Notes:
* I had to put `attach: false` in the StopwatchNode listeners.
* I had to specify the `targetNode` since `event.trail.lastNode` is a child, not the thing we typically want to move to the front.  Is there a way to automatically find the node a listener is registered on?  Should `targetNode` be an option instead of a required parameter?
* I added another listener when the StopwatchNode is dragged out of the toolbox.  I couldn't find a way to make this automatically part of the `MoveToFrontListener`.

Here is a proposed plan for next steps:
1. @pixelzoom can you please review this patch?
2. If it seems reasonable, let's commit to master, mark blocks-publication and make a chip-away for converting sims to use this pattern.

Thoughts?

**Comment by pixelzoom** [comment]
Yes, your patch is exactly what I had in mind. 

The sims that `import StopwatchNode` are:

- [ ] CL:B
- [ ] CCK-common
- [ ] collision-lab
- [ ] energy-skate-park
- [ ] gas-properties
- [ ] gravity-and-orbits
- [ ] masses-and-springs
- [ ] pendulum-lab
- [ ] wave-interference
- [ ] wave-on-a-string

If we add a chip-away issue, I think it will be a long time before this gets done.  How about if we do a PSA at developer meeting that we're going to do this, and I take a stab at applying your patch and fixing up the usages?

**Comment by mattpen** [comment]
Dev Meeting 3/11: @pixelzoom announced this as a PSA and will continue with applying the patch.

**Comment by chrisklus** [comment]
Oops, I misunderstood and thought this was a chip-away. Unassigning.

**Comment by pixelzoom** [comment]
Oh boy, this went sideways quickly. Luckily I tried this in GasProperties first, and hit the problem immediately. 

Here's the DragListener in CollisionCounterNode:

```js
    // Move to front on pointer down, anywhere on this Node, including interactive subcomponents.
    // This needs to be a DragListener so that touchSnag works.
    this.addInputListener( new DragListener( {
      attach: false, // so that this DragListener won't be ignored
      start: () => {
        listboxParent.moveToFront();
      },
      tandem: options.tandem.createTandem( 'moveToFrontListener' )
    } ) );
```

Note the comment that says "This needs to be a DragListener so that touchSnag works."

So `MoveToFrontListener extends PressListener` is not going to do it.  When the user drags a Node using touchSnag, the Node will not move to front. 

Options:
(1) `MoveToFrontListener extends DragListener` and handle this in `options.start`. Seems kind of heavyweight, and I'm wondering if there would be problems having multiple DragListeners on a Node. But if it works...
(2) Add a `moveToFront` option to DragListener, since this is something that seems like a common need - move something to the front when dragging starts.
(3) Investigate supporting `touchSnag` in listeners other than DragListener. Is it relevant for PressListener? It seems to be in this case. This would involved @jonathanolson.

And another option/issue that ocurred to me:
It seems appropriate to remove not only "move to front", but the entire DragListener from StopwatchNode. There seems to be an assumption that all stop watches are movable, and that's not the case. RulerNode, ProbeNode,... most UI components are not responsibile for their dragging. Why should StopwatchNode have dragging baked in?

@samreid thoughts?

**Comment by pixelzoom** [comment]
I'm liking this option the best:
> (2) Add a moveToFront option to DragListener, since this is something that seems like a common need - move something to the front when dragging starts.

@jonathanolson @samreid thoughts?

**Comment by jonathanolson** [comment]
I want to see if I'm understanding the desired behavior correctly:

- The StopwatchNode has two buttons with listeners
- The client usually adds a listener (usually a DragListener) on the `dragTarget` of the StopwatchNode
- We want to add a listener that activates moveToFront() when ANY of the above listeners enter their press state, but not otherwise

How do we want to handle when the listener on the StopwatchNode isn't a PressListener? When it's not a DragListener?

It seems like the above approach is "create a PressListener on the entire composite", but that fails because sometimes we touch-snag (and maybe sometimes we don't?)

My first thought is:

1. Add an emitter on StopwatchNode that fires when either of its buttons are activated.
2. The client, if it wants, can moveToFront when that emitter fires
3. The client, if it wants, can moveToFront when the listener it created starts.

**Comment by jonathanolson** [comment]
> Investigate supporting touchSnag in listeners other than DragListener. Is it relevant for PressListener? It seems to be in this case.

That seems like it doesn't solve the above problems. The up-front cost to move touchSnag to PressListener isn't too much, but it feels weird. touchSnag would be an option on FireListener then? I'm not necessarily against it, but I still see problems with using that approach.

**Comment by jonathanolson** [comment]
The general idea of MoveToFrontListener sounds really tricky. It sounds like we're trying to duplicate the conditions that would activate other listeners. I generally don't like the idea of trying to coordinate multiple listeners so they both "press" and "release" at the same time.

What about when one is disabled? What if the user adds specific conditions to one for it to be able to start a press? In this specific case, what happens when you add one on the `targetNode` and one on the composite? Sounds really tricky. For instance, in this case, what happens when the entire StopwatchNode is disabled?

**Comment by pixelzoom** [comment]
I think we need a subgroup meeting for this: me, @samreid, and @jonathanolson.

Summary:  `moveToFront` and DragListener is currently baked into StopwatchNode  I'm recommending that `moveToFront` and DragListeners don't belong in common-code components like RulerNode, StopwatchNode, etc.   But we have a need for common behavior related to `moveToFront` and DragListeners (particularly for "tools") and we don't want to duplicate in every client.  Additionally `moveToFront` needs to occur on touch snag.  So what is the best way to proceed?

**Comment by pixelzoom** [comment]
Here's a list (possibly incomplete) of scenery-phet "tools"  that might be draggable, but are not necessarily draggable.

* ConductivityTesterNode
* EyeDropperNode
* GaugeNode (and ValueGaugeNode)
* MagnifyingGlassNode
* MeasuringTapeNode
* ProbeNode
* ProtractorNode
* RulerNode
* StopwatchNode
* ThermometerNode

**Comment by samreid** [comment]
After discussion today (@pixelzoom @jonathanolson @samreid):

* Add the 2nd DragListener from CollisionCounterNode to StopwatchNode, so it will move to front on picking (even when picking a disabled button).  @pixelzoom volunteered to take a look at this.
* We discussed factoring out https://github.com/phetsims/gas-properties/blob/5616f3b052f6592ce0575a7aefe42954b16f529b/js/common/view/CollisionCounterNode.js#L199-L228, but it is unclear how this could be done efficiently
* We discussed moving some DragListeners to be extrinsic to some nodes (say, dragging StopwatchNode could be extrinsic, whereas dragging a MeasuringTapeNode could be intrinsic).
* We agreed it is unfortunate that (a) behavior of tools is inconsistent and (b) it is difficult to use one listener or one listener type for all tools within a sim, since some common code components want to handle it themselves.

**Comment by pixelzoom** [comment]
More notes from today's discussion....

StopwatchNode options `visibleBoundsProperty` should be renamed.  It probably started off as `visibleBoundsProperty` from Sim (the visible bounds of the browser window),  but it can really be any drag bounds. 

The 2nd DragListener in CollisionCounterNode, which I'll add to StopwatchNode, is:

```js
    // Move to front on pointer down, anywhere on this Node, including interactive subcomponents.
    // This needs to be a DragListener so that touchSnag works.
    this.addInputListener( new DragListener( {
      attach: false, // so that this DragListener won't be ignored
      start: () => {
        this.moveToFront();
        listboxParent.moveToFront();
      },
      tandem: options.tandem.createTandem( 'moveToFrontListener' )
    } ) );
```

We discussed other ways of doing this (Emitter, ...) and decided that this was the lesser evil for now.  We noted that the `touchSnag` aspect of this 2nd DragListener is not currently working, and that will be addressed in phetsims/scenery#1177.

**Comment by pixelzoom** [comment]
Other cleanup to address:

- [ ] This field is unnecessary, not used outside of the constructor or by clients. Code would be clearer if `this.dragTarget` were replaced with `backgroundNode`.

```
    // @public (read-only) - Target for drag listeners
    this.dragTarget = backgroundNode;
```

**Comment by pixelzoom** [comment]
Having 2 DragListeners was resulting in undesirable behavior with touchSnag, see https://github.com/phetsims/scenery/issues/1177#issuecomment-801301269. So after discussion with @jonathanolson, we revised the approach used for the 2nd input listener in the 2 commits directly above.

**Comment by pixelzoom** [comment]
Having 2 DragListeners was resulting in undesirable behavior with touchSnag, see https://github.com/phetsims/scenery/issues/1177#issuecomment-801301269. So after discussion with @jonathanolson, we revised the approach used for the 2nd input listener in f1e6c0c.  Here's the most important bit:


```diff
-      this.addInputListener( new DragListener( {
-         attach: false, // so that this DragListener won't be ignored
-         start: () => this.moveToFront(),
-         tandem: options.tandem.createTandem( 'moveToFrontListener' )
-       } ) );
+       this.addInputListener( {
+         down: () => this.moveToFront()
 +      } );
```


Note that this change also means that the 2nd listener is not instrumented, and this will break PhET-iO APIs for sims that have a StopwatchNode.  @samreid what do we need to do about that?

**Comment by pixelzoom** [comment]
Work is done, @samreid please review.     Also note the question about PhET-iO API breakage in https://github.com/phetsims/scenery-phet/issues/659#issuecomment-801332224.

**Comment by jonathanolson** [comment]
> This field is unnecessary, not used outside of the constructor or by clients. Code would be clearer if this.dragTarget were replaced with backgroundNode

Is that a convention that we're going to use `backgroundNode` for all of these cases (components with undraggable sections)? Many times it doesn't seem like it would be a background, so the name was explicitly chosen so that its purpose was clear.

**Comment by pixelzoom** [comment]
I don't care what's it's called.  What I was objecting to was creating a public field that is currently unnecessary and unused (and therefore untested).  But I'll restore it.

**Comment by pixelzoom** [comment]
`this.dragTarget` restored in the above commit.  I'll also point out that input listener APIs (including DragListener) use `targetNode`.  So `dragTarget` is inconsistent.  But I feel like I'm touching things that I shouldn't be, and will leave these rough edges to be polished by the authors.

**Comment by samreid** [comment]
I discussed `this.dragTarget` with @jonathanolson and we agreed that `dragTarget` will be a nice name if we ever need a public API for it, but at the moment there is no need for it.  I'll cherry pick: 40b6e47328fcad90583c96e70c1e0b09818ca4ba

**Comment by samreid** [comment]
> Note that this change also means that the 2nd listener is not instrumented, and this will break PhET-iO APIs for sims that have a StopwatchNode. @samreid what do we need to do about that?

I don't think we need to instrument the move to front listener, nor do we need to maintain backward compatibility with sims that exposed that in a prior version.

I tested the changes in Waves Intro with mouse and touch, over normal and disabled components and the behavior seemed reasonable.  I reviewed the code changes in StopwatchNode and skimmed the usage sites and did not have any recommendations.  I think this issue can be closed.
